README.md
Facebook Mod Bot
================

<p align="center">
  <img src="http://i.imgur.com/xqpBTlg.png" alt="Sublet Bot" height="250" width="250"/>
</p>

**[Changelog](https://github.com/pandanomic/FB_Mod_Bot/blob/master/CHANGELOG.md)**

This is a facebook bot that I wrote to moderate a group I admin, essentially keeping some order and upholding some requirements per the group rules. The main purpose is to crack down on vague/uninformative posts.

Anyone is welcome to use this code and/or repurpose it for their own use. All I ask is that you give me credit somewhere :)

The bot is written in Python, using the [facebook-sdk](https://github.com/pythonforfacebook/facebook-sdk) API wrapper for Facebook's graph API. To make the bot functional and separate from my personal account, I created a dedicated Facebook account for my bot (literally called "Sublets Bot"), registered it as a developer, created an app, granted that same app full permissions to the account, and finally use that access token in the bot itself to authenticate with Facebook's Graph API.

  * *Why make a dedicated app instead of just creating an access token? Because only an app (with an app ID and secret key) can programatically request an extended access token. If I don't do that, I'd have to manually generate a new one every two hours.* 

The group is a sublets/roommate finding group for students at my university, where people can post sublet offerings, say they're looking for one, and also seek out other potential roommates.

### The bot has three major validation checks:
* Pricing reference
  * There must be some sort of pricing referenced, either with `$` signs or by saying `____/month` or `____ per month`
* Minimum length or craigslist link
  * If their post is under 200 characters, doesn't have a link to a craiglist ad, and isn't a parking offering, then they need to edit and include more details.
* Proper tag
  * For easy searching, we require them to prepend their post with `[LOOKING]`, `[OFFERING]`, `[ROOMING]`, or `[PARKING]`.
  * The bot is purposefully a little fuzzy on this due to the large number of people that can't find the bracket keys on their keyboards ಠ_ಠ

If any of those checks fail, it comments on the user notifying them of the problem(s) and specifies what it(they) are. Upon warning, it caches the post ID and time, giving them 24 hours to fix their post. On later runs, if a previously warned post is now valid, it removes the warning comment and the post ID from the cache. ~~I'd send a message thanking them too, but Facebook's API doesn't allow apps to send messages.~~ Yay XMPP

### TODO
* ~~Prepare for Heroku hosting~~ Done!
* Automatically delete posts
* Improve documentation

### Caveats
* Facebook's API doesn't allow you to delete anything that the app itself didn't create. 
  * Therefore, you'll have to delete any other posts manually. Locally I have it set to just open the post-to-be-deleted in a new browser tab. On Heroku, I'll set up some form of sending those links back to me.
* Depending on how many posts you request at once (the `LIMIT` option in FQL queries), Facebook's API periodically just crashes and returns an internal server error. Fortunately this doesn't seem to happen unless you ask for extremely high numbers.


## License

     The MIT License (MIT)

     Copyright (c) 2014 Henri Sweers

     Permission is hereby granted, free of charge, to any person obtaining a copy of
     this software and associated documentation files (the "Software"), to deal in
     the Software without restriction, including without limitation the rights to
     use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
     the Software, and to permit persons to whom the Software is furnished to do so,
     subject to the following conditions:

     The above copyright notice and this permission notice shall be included in all
     copies or substantial portions of the Software.

     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
     IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
     FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
     COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
     IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
     CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
FB Mod Bot Changelog
====================

### 02/08/14
* Created changelog
* Added `[PARKING]` tag checking

facebook_sublet_group_bot/test.py
__author__ = 'pandanomic'

import unittest


class TestTagValidity(unittest.TestCase):
    def setUp(self):
        self.test_tags = ["Looking", "Rooming", "Offering", "Parking"]
        self.junk = """here's some other text because yeah more text to
            to illustrate lots more text here in the rest of the post"""

    def test_regular(self):
        from fb_bot import check_tag_validity
        for tag in self.test_tags:
            self.assertTrue(check_tag_validity("(" + tag + ")"))
            self.assertTrue(check_tag_validity("{" + tag + "}"))
            self.assertTrue(check_tag_validity("[" + tag + "]"))
            self.assertTrue(check_tag_validity("(" + tag + "]"))
            self.assertTrue(check_tag_validity("{" + tag + "]"))
            self.assertTrue(check_tag_validity("(" + tag + "}"))
            self.assertTrue(check_tag_validity("{" + tag + ")"))
            self.assertFalse(check_tag_validity(tag + ")"))
            self.assertFalse(check_tag_validity("{" + tag))
            self.assertFalse(check_tag_validity(tag))

    def test_misc(self):
        from fb_bot import check_tag_validity
        for tag in self.test_tags:
            self.assertTrue(check_tag_validity("(" + tag + ") sometjunk"))
            self.assertFalse(check_tag_validity("{" + tag + "}" + self.junk))
            self.assertFalse(check_tag_validity("dsflkj{" + tag.lower() + ")"))
            self.assertFalse(check_tag_validity("dsflkj {" + tag.lower() + ")"))
            self.assertTrue(check_tag_validity("-(" + tag + ")"))
            self.assertTrue(check_tag_validity("*(" + tag + ")"))
            self.assertTrue(check_tag_validity("* (" + tag + ")"))
            self.assertTrue(check_tag_validity(" (" + tag + ")"))
            self.assertTrue(check_tag_validity("(" + tag + "):"))
            self.assertTrue(check_tag_validity("(" + tag + ") :"))

    def test_parking(self):
        from fb_bot import check_for_parking_tag
        tag = "parking"
        self.assertTrue(check_for_parking_tag("(" + tag + ")"))
        self.assertTrue(check_for_parking_tag("{" + tag + "}"))
        self.assertTrue(check_for_parking_tag("[" + tag + "]"))
        self.assertTrue(check_for_parking_tag("(" + tag + "]"))
        self.assertTrue(check_for_parking_tag("{" + tag + "]"))
        self.assertTrue(check_for_parking_tag("(" + tag + "}"))
        self.assertTrue(check_for_parking_tag("{" + tag + ")"))
        self.assertFalse(check_for_parking_tag(tag + ")"))
        self.assertFalse(check_for_parking_tag("{" + tag))
        self.assertFalse(check_for_parking_tag(tag))
        self.assertTrue(check_for_parking_tag("(" + tag + ") sometjunk"))
        self.assertFalse(check_for_parking_tag("{" + tag + "}" + self.junk))
        self.assertFalse(check_for_parking_tag("dsflkj{" + tag.lower() + ")"))
        self.assertFalse(check_for_parking_tag("dsflkj {" + tag.lower() + ")"))
        self.assertTrue(check_for_parking_tag("-(" + tag + ")"))
        self.assertTrue(check_for_parking_tag("*(" + tag + ")"))
        self.assertTrue(check_for_parking_tag("* (" + tag + ")"))
        self.assertTrue(check_for_parking_tag(" (" + tag + ")"))
        self.assertTrue(check_for_parking_tag("(" + tag + "):"))
        self.assertTrue(check_for_parking_tag("(" + tag + ") :"))


class TestPriceValidity(unittest.TestCase):
    def setUp(self):
        self.junk = """here's some other text because yeah more text to
            to illustrate lots more text here in the rest of the post"""

    def test_regular_month(self):
        from fb_bot import check_price_validity
        self.assertTrue(check_price_validity("Blah blah $ blah"))
        self.assertTrue(check_price_validity("Blah blah 300 per month"))
        self.assertTrue(check_price_validity("Blah blah 300/month"))
        self.assertTrue(check_price_validity("Blah blah 300 / month"))
        self.assertTrue(check_price_validity("Blah blah 300 /month"))
        self.assertTrue(check_price_validity("Blah blah 300/ month"))
        self.assertTrue(check_price_validity("Blah blah 300 per/month"))
        self.assertTrue(check_price_validity("Blah blah 300 a month"))
        self.assertTrue(check_price_validity("Blah blah 300 a/month"))

    def test_embedded_month(self):
        from fb_bot import check_price_validity
        self.assertTrue(check_price_validity("Blah $ blah" + self.junk))
        self.assertTrue(check_price_validity("Blah 300 per month" + self.junk))
        self.assertTrue(check_price_validity("Blah 300/month" + self.junk))
        self.assertTrue(check_price_validity("Blah 300 / month" + self.junk))
        self.assertTrue(check_price_validity("Blah 300 /month" + self.junk))
        self.assertTrue(check_price_validity("Blah 300/ month" + self.junk))
        self.assertTrue(check_price_validity("Blah 300 per/month" + self.junk))
        self.assertTrue(check_price_validity("Blah 300 a month" + self.junk))
        self.assertTrue(check_price_validity("Blah 300 a/month" + self.junk))

    def test_misc_month(self):
        from fb_bot import check_price_validity
        self.assertTrue(check_price_validity("Blah$ blah" + self.junk))
        self.assertTrue(check_price_validity("Blah300 per month" + self.junk))
        self.assertTrue(check_price_validity("Blah blah 300permonth"))
        self.assertTrue(check_price_validity("Blah blah300/monthblah"))
        self.assertTrue(check_price_validity("Blah blah 300amonth"))

    def test_regular_mon(self):
        from fb_bot import check_price_validity
        self.assertTrue(check_price_validity("Blah blah $ blah"))
        self.assertTrue(check_price_validity("Blah blah 300 per mon"))
        self.assertTrue(check_price_validity("Blah blah 300/mon"))
        self.assertTrue(check_price_validity("Blah blah 300 / mon"))
        self.assertTrue(check_price_validity("Blah blah 300 /mon"))
        self.assertTrue(check_price_validity("Blah blah 300/ mon"))
        self.assertTrue(check_price_validity("Blah blah 300 per/mon"))
        self.assertTrue(check_price_validity("Blah blah 300 a mon"))
        self.assertTrue(check_price_validity("Blah blah 300 a/mon"))

    def test_embedded_mon(self):
        from fb_bot import check_price_validity
        self.assertTrue(check_price_validity("Blah $ blah" + self.junk))
        self.assertTrue(check_price_validity("Blah 300 per mon" + self.junk))
        self.assertTrue(check_price_validity("Blah 300/mon" + self.junk))
        self.assertTrue(check_price_validity("Blah 300 / mon" + self.junk))
        self.assertTrue(check_price_validity("Blah 300 /mon" + self.junk))
        self.assertTrue(check_price_validity("Blah 300/ mon" + self.junk))
        self.assertTrue(check_price_validity("Blah 300 per/mon" + self.junk))
        self.assertTrue(check_price_validity("Blah 300 a mon" + self.junk))
        self.assertTrue(check_price_validity("Blah 300 a/mon" + self.junk))

    def test_misc_mon(self):
        from fb_bot import check_price_validity
        self.assertTrue(check_price_validity("Blah$ blah" + self.junk))
        self.assertTrue(check_price_validity("Blah300 per mon" + self.junk))
        self.assertTrue(check_price_validity("Blah blah 300permon"))
        self.assertTrue(check_price_validity("Blah blah300/monblah"))
        self.assertTrue(check_price_validity("Blah blah 300amon"))

    def test_regular_mo(self):
        from fb_bot import check_price_validity
        self.assertTrue(check_price_validity("Blah blah $ blah"))
        self.assertTrue(check_price_validity("Blah blah 300 per mo"))
        self.assertTrue(check_price_validity("Blah blah 300/mo"))
        self.assertTrue(check_price_validity("Blah blah 300 / mo"))
        self.assertTrue(check_price_validity("Blah blah 300 /mo"))
        self.assertTrue(check_price_validity("Blah blah 300/ mo"))
        self.assertTrue(check_price_validity("Blah blah 300 per/mo"))
        self.assertTrue(check_price_validity("Blah blah 300 a mo"))
        self.assertTrue(check_price_validity("Blah blah 300 a/mo"))

    def test_embedded_mo(self):
        from fb_bot import check_price_validity
        self.assertTrue(check_price_validity("Blah $ blah" + self.junk))
        self.assertTrue(check_price_validity("Blah 300 per mo" + self.junk))
        self.assertTrue(check_price_validity("Blah 300/mo" + self.junk))
        self.assertTrue(check_price_validity("Blah 300 / mo" + self.junk))
        self.assertTrue(check_price_validity("Blah 300 /mo" + self.junk))
        self.assertTrue(check_price_validity("Blah 300/ mo" + self.junk))
        self.assertTrue(check_price_validity("Blah 300 per/mo" + self.junk))
        self.assertTrue(check_price_validity("Blah 300 a mo" + self.junk))
        self.assertTrue(check_price_validity("Blah 300 a/mo" + self.junk))

    def test_misc_mo(self):
        from fb_bot import check_price_validity
        self.assertTrue(check_price_validity("Blah$ blah" + self.junk))
        self.assertTrue(check_price_validity("Blah300 per mo" + self.junk))
        self.assertTrue(check_price_validity("Blah blah 300permo"))
        self.assertTrue(check_price_validity("Blah blah300/moblah"))
        self.assertTrue(check_price_validity("Blah blah 300amo"))


if __name__ == '__main__':
    unittest.main()

delete.py
from selenium import webdriver
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.support.wait import WebDriverWait
import contextlib
import time


def delete_post(usr, pwd, post):
    print "STARTING DELETE"
    sleep_time = 0

    with contextlib.closing(webdriver.PhantomJS()) as driver:
        # or you can use Firefox()
        # or you can use Chrome(executable_path="/usr/bin/chromedriver")
        print "--Opening facebook..."
        driver.get("http://www.facebook.org")
        time.sleep(sleep_time)
        assert "Facebook" in driver.title

        print "--Logging in..."
        elem = WebDriverWait(driver, 10).until(lambda driver : driver.find_element_by_id("email"))
        time.sleep(sleep_time)
        elem.send_keys(usr)
        time.sleep(sleep_time)
        elem = WebDriverWait(driver, 10).until(lambda driver : driver.find_element_by_id("pass"))
        time.sleep(sleep_time)
        elem.send_keys(pwd)
        time.sleep(sleep_time)
        elem.send_keys(Keys.RETURN)
        time.sleep(sleep_time)

        print "--Retrieving post..."
        driver.get(post)
        time.sleep(sleep_time)

        print "--Expanding arrow options..."
        elem = WebDriverWait(driver, 10).until(lambda driver : driver.find_element_by_xpath("//a[contains(@class,'_5pbj _p')]"))
        time.sleep(sleep_time)
        elem.click()
        time.sleep(sleep_time)

        print "--Finding delete option..."
        elem = WebDriverWait(driver, 10).until(lambda driver : driver.find_element_by_xpath("//div[contains(@class,'_54ng')]"))
        time.sleep(sleep_time)
        links = WebDriverWait(driver, 10).until(lambda driver : elem.find_elements_by_xpath('.//a'))
        time.sleep(sleep_time)

        print "--Clicking delete..."
        links[2].click()
        time.sleep(sleep_time)

        print "--Finding delete confirmation 'enter'..."
        elem = WebDriverWait(driver, 10).until(lambda driver : driver.find_element_by_css_selector("button._42fu:nth-child(1)"))
        time.sleep(sleep_time)

        print "--Deleting..."
        elem.click()

        print "Deleted"

facebook_sublet_group_bot/fb_bot.py
#!/usr/bin/env python

import os
import cPickle as pickle
import webbrowser
import datetime
import facebook
import time
import subprocess
import sys
import re
from fbxmpp import SendMsgBot
from delete_post import delete_post

__author__ = 'Henri Sweers'

# Color class, used for colors in terminal
class Color:
    PURPLE = '\033[95m'
    CYAN = '\033[96m'
    DARKCYAN = '\033[36m'
    BLUE = '\033[94m'
    GREEN = '\033[92m'
    YELLOW = '\033[93m'
    RED = '\033[91m'
    BOLD = '\033[1m'
    UNDERLINE = '\033[4m'
    END = '\033[0m'


# Log method. If there's a color argument, it'll stick that in first
def log(message, *colorargs):
    if len(colorargs) > 0:
        print colorargs[0] + message + Color.END
    else:
        print message

############## Global vars ##############

# 24 hours, in seconds
time_limit = 86400

# Pickle cache file caching warned posts
warned_db = "fb_subs_cache"

# Pickle cache file caching valid posts
valid_db = "fb_subs_valid_cache"

# Pickle cache file for properties
prop_file = "login_prop"

# Boolean for key extensions
extend_key = False

# Boolean for checking heroku
running_on_heroku = False


# Junk method that I use for testing stuff periodically
def test():
    log('Test', Color.PURPLE)


# Method for sending messages, adapted from here: http://goo.gl/oV5KtZ
def send_message(recipient, botid, message, api_key, access_token):
    # The "From" Facebook ID
    jid = botid + '@chat.facebook.com'

    # The "Recipient" Facebook ID, with a hyphen for some reason
    to = '-' + recipient + '@chat.facebook.com'

    xmpp = SendMsgBot(jid, to, unicode(message))

    xmpp.credentials['api_key'] = api_key
    xmpp.credentials['access_token'] = access_token

    if xmpp.connect(('chat.facebook.com', 5222)):
        xmpp.process(block=True)
        log('----Message sent', Color.GREEN)
    else:
        log("----Unable to connect, message sending fail", Color.RED)


# Use this method to set new vals for props, such as on your first run
def set_new_props():
    saved_dict = load_properties()

    #### Uncomment lines below as needed to manually set stuff ####

    #### These are strings
    # saved_dict['sublets_oauth_access_token'] = "put-auth-token-here"
    # saved_dict['sublets_api_id'] = "put-app-id-here"
    # saved_dict['sublets_secret_key'] = "put-secret-key-here"
    # saved_dict['access_token_expiration'] = "put-access-token-expiration-here"
    # saved_dict['group_id'] = "put-group-id-here"

    #### These are ints
    # saved_dict['bot_id'] = put-bot-id-here
    # saved_dict['ignored_post_ids'].append(<id_num>)
    # saved_dict['ignore_source_ids'].append(<id_num>)
    # saved_dict['admin_ids'].append(<id_num>)

    #### You can do other stuff too, the above are just examples ####

    save_properties(saved_dict)


# Method for initializing your prop values
def init_props():
    test_dict = {'sublets_oauth_access_token': "put-auth-token-here",
                 'sublets_api_id': "put-app-id-here",
                 'sublets_secret_key': "put-secret-key-here",
                 'group_id': 'put-group-id-here',
                 'bot_id': -1,
                 'ignored_post_ids': [],
                 'ignore_source_ids': [],
                 'admin_ids': []}
    save_properties(test_dict)
    saved_dict = load_properties()
    assert test_dict == saved_dict


# Method for saving (with pickle) your prop values
def save_properties(data):
    if running_on_heroku:
        mc.set('props', data)
    else:
        with open(prop_file, 'w+') as login_prop_file:
            pickle.dump(data, login_prop_file)


# Method for loading (with pickle) your prop values
def load_properties():
    if running_on_heroku:
        obj = mc.get('props')
        if not obj:
            return {}
        else:
            return obj
    else:
        if os.path.isfile(prop_file):
            with open(prop_file, 'r+') as login_prop_file:
                data = pickle.load(login_prop_file)
                return data
        else:
            sys.exit("No prop file found")


# Method for loading a cache. Either returns cached values or original data
def load_cache(cachename, data):
    if running_on_heroku:
        if running_on_heroku:
            obj = mc.get(cachename)
            if not obj:
                return data
            else:
                return obj
    else:
        if os.path.isfile(cachename):
            with open(cachename, 'r+') as f:

                # If the file isn't at its end or empty
                if f.tell() != os.fstat(f.fileno()).st_size:
                    return pickle.load(f)
        else:
            log("--No cache file found, a new one will be created", Color.BLUE)
            return data


# Method for saving to cache
def save_cache(cachename, data):
    if running_on_heroku:
        mc.set(cachename, data)
    else:
        with open(cachename, 'w+') as f:
            pickle.dump(data, f)


# Nifty method for sending notifications on my mac when it's done
def notify_mac():
    if sys.platform == "darwin":
        try:
            subprocess.call(
                ["terminal-notifier", "-message", "Done", "-title", "FB_Bot",
                 "-sound", "default"])
        except OSError:
            print "If you have terminal-notifier, this would be a notification"


# Method for checking tag validity
def check_tag_validity(message_text):
    p = re.compile(
        "^(-|\*| )*([\(\[\{])((looking)|(rooming)|(offering)|(parking))([\)\]\}])(:)?(\s|$)",
        re.IGNORECASE)

    if re.match(p, message_text):
        return True
    else:
        return False


# Method for checking if pricing reference is there
def check_price_validity(message_text):
    p = re.compile(
        "(\$)|((\d)+( )?((per)|(/)|(a))( )?(/)?((month)|(mon)|(mo))(\s)?)",
        re.IGNORECASE)

    if re.search(p, message_text) is not None:
        return True
    else:
        return False


# Checking if there's a parking tag
def check_for_parking_tag(message_text):
    p = re.compile(
        "^(-|\*| )*([\(\[\{])(parking)([\)\]\}])(:)?(\s|$)",
        re.IGNORECASE)

    if re.search(p, message_text):
        return True
    else:
        return False


# Method for extending access token
def extend_access_token(graph, now_time, saved_props, sublets_api_id,
                        sublets_secret_key):
    log("Extending access token", Color.BOLD)
    result = graph.extend_access_token(sublets_api_id, sublets_secret_key)
    new_token = result['access_token']
    new_time = int(result['expires']) + now_time
    saved_props['sublets_oauth_access_token'] = new_token
    saved_props['access_token_expiration'] = new_time
    log("Token extended", Color.BOLD)


# Method for retrieving user ID's of admins in group, ignoring bot ID
def retrieve_admin_ids(group_id, bot_id, auth_token):

    # Retrieve the uids via FQL query
    graph = facebook.GraphAPI(auth_token)
    admins_query = \
        "SELECT uid FROM group_member WHERE gid=" + group_id + " AND" + \
        " administrator AND NOT (uid = " + str(bot_id) + ")"
    admins = graph.fql(query=admins_query)

    # Parse out the uids from the response
    admins_list = [admin['uid'] for admin in admins]

    # Update the admin_ids in our properties
    saved_props = load_properties()
    saved_props['admin_ids'] = admins_list
    save_properties(saved_props)

    return admins_list


# Extracted logic for messaging admins a message
def message_admins(message, auth_token, app_id, bot_id, group_id):
    for admin in retrieve_admin_ids(group_id, bot_id, auth_token):
        send_message(str(admin), str(bot_id), message, app_id, auth_token)


# Main runner method
def sub_group():
    # Load the properties
    saved_props = load_properties()

    # Access token
    sublets_oauth_access_token = saved_props['sublets_oauth_access_token']

    # Access token expiration
    access_token_expiration = saved_props['access_token_expiration']

    # API App ID
    sublets_api_id = saved_props['sublets_api_id']

    # API App secret key
    sublets_secret_key = saved_props['sublets_secret_key']

    # List of posts to ignore
    ignored_post_ids = saved_props['ignored_post_ids']

    # List of people to ignore
    ignore_source_ids = saved_props['ignore_source_ids']

    # ID of the FB group
    group_id = saved_props['group_id']

    # User ID of the bot
    bot_id = saved_props['bot_id']

    # IDs of admins (unused right now, might remove later)
    admin_ids = saved_props['admin_ids']

    # FQL query for the group
    group_query = "SELECT post_id, message, actor_id FROM stream WHERE " + \
                  "source_id=" + group_id + " LIMIT 50"

    # Get current time
    now_time = time.time()

    # For logging purposes
    log("CURRENT CST TIMESTAMP: " + datetime.datetime.fromtimestamp(
        now_time - 21600).strftime('%Y-%m-%d %H:%M:%S'), Color.UNDERLINE)

    # Make sure the access token is still valid
    if access_token_expiration < now_time:
        sys.exit("API Token is expired")

    # Warn if the token's expiring soon
    if access_token_expiration - now_time < 604800:
        log("Warning - access token expires in less than a week", Color.RED)
        log("-- Expires on " + datetime.datetime.fromtimestamp(
            access_token_expiration).strftime('%Y-%m-%d %H:%M:%S'))

        # If you want it to automatically when it's close to exp.
        global extend_key
        extend_key = True

    # Log in, try to get posts
    graph = facebook.GraphAPI(sublets_oauth_access_token)

    # Extend the access token, default is ~2 months from current date
    if extend_key:
        extend_access_token(graph, now_time, saved_props, sublets_api_id,
                            sublets_secret_key)

    # Make our first request, get the group posts
    group_posts = graph.fql(query=group_query)

    # Load the pickled cache of previously warned posts
    already_warned = dict()
    log("Loading warned cache", Color.BOLD)
    already_warned = load_cache(warned_db, already_warned)
    log('--Loading cache size: ' + str(len(already_warned)), Color.BOLD)

    # Load the pickled cache of valid posts
    valid_posts = []
    log("Checking valid cache.", Color.BOLD)
    valid_posts = load_cache(valid_db, valid_posts)
    log('--Valid cache size: ' + str(len(valid_posts)), Color.BOLD)

    # Loop over retrieved posts
    for post in group_posts:

        # Important data received
        post_message = post['message']      # Content of the post
        post_id = post['post_id']           # Unique ID of the post

        # Unique ID of the person that posted it
        actor_id = post['actor_id']

        # Ignore mods and certain posts
        if post_id in ignored_post_ids or actor_id in ignore_source_ids or \
                post_id in valid_posts:
            # log('\n--Ignored post: ' + post_id, Color.BLUE)
            continue

        # Data to use
        post_comment = "(This is an automated comment)\n\nIt looks like" + \
                       " your post has a few issues:\n"

        # Boolean for tracking if the post is valid
        valid_post = True

        # Log the message details
        log("\n" + post_message[0:75].replace('\n', "") + "...\n--POST ID: " +
            str(post_id) + "\n--ACTOR ID: " + str(actor_id))

        # Check for pricing
        if not check_price_validity(post_message):
            valid_post = False
            post_comment += "- Your post doesn't seem to mention pricing" + \
                " (no $ signs, <number>/month), or \"per month\")\n"
            log('----$', Color.BLUE)

        # Check for tag validity
        if not check_tag_validity(post_message):
            valid_post = False
            post_comment += \
                "- Your post appears to be missing a proper tag at the" + \
                " front ([LOOKING], [ROOMING], or [OFFERING])\n"
            log('----Tag', Color.BLUE)

        # Check post length.
        # Allow short ones if there's a craigslist link or parking
        if len(post_message) < 200 and \
                "craigslist" not in post_message.lower() \
                and not check_for_parking_tag(post_message):
            valid_post = False
            post_comment += \
                "- Your post is a little short (<200 chars). Please give" + \
                " more information (what you're looking for, details," + \
                " preferences, etc.)\n"
            log('----Length', Color.BLUE)

        # Not a valid post
        if not valid_post:

            # If already warned, delete if it's been more than 24 hours, ignore
            # if it's been less
            if post_id in already_warned:

                # Invalid, past 24 hour grace period
                if time_limit < now_time - already_warned[post_id]:
                    log('--Delete: ' + post_id, Color.RED)
                    url = "http://www.facebook.com/" + post_id

                    # If local, just open it in a new browser tab
                    if not running_on_heroku:
                        webbrowser.open_new_tab(url)
                        del already_warned[post_id]

                    # Otherwise message the admins the URL of the post to delete
                    else:
                        try:
                            delete_post(url)
                            message_admins(
                                "Please make sure this is gone: " + url,
                                sublets_oauth_access_token,
                                sublets_api_id, bot_id, group_id)
                        except Exception as e:
                            message_admins(
                                "Something went wrong, check heroku logs",
                                sublets_oauth_access_token,
                                sublets_api_id, bot_id, group_id)
                            log(e.message, Color.RED)

                # Invalid but they still have time
                else:
                    time_delta = time_limit - (now_time -
                                               already_warned[post_id])
                    m, s = divmod(time_delta, 60)
                    h, m = divmod(m, 60)
                    log_message = '--Invalid, but still have '
                    if h > 0:
                        log_message += '%d hours and ' % h
                    log_message += '%02d minutes' % m
                    log(log_message, Color.RED)
                continue

            # Comment with a warning and cache the post
            else:

                # First check to make sure we haven't warned them before
                # by searching comments for bot comment
                previously_commented = False
                comments_query = "SELECT fromid, id, time FROM comment" + \
                                 " WHERE post_id=\"" + str(post_id) + "\""
                comments = graph.fql(comments_query)
                for comment in comments:

                    # Found a comment from the bot
                    if comment['fromid'] == bot_id:
                        log('--Previously warned')
                        log('----caching')
                        previously_commented = True
                        already_warned[post_id] = comment['time']
                        break

                # Comment if no previous comment
                if not previously_commented:
                    # Comment to post for warning
                    post_comment += \
                        "\nPlease edit your post and fix the above within 24" +\
                        " hours, or else your post will be deleted per the" + \
                        " group rules. Thanks!\n\n(If you think this was a" + \
                        " mistake don't hesitate to message me)"

                    graph.put_object(
                        post['post_id'], "comments", message=post_comment)
                    # Save
                    already_warned[post_id] = now_time
                    log('--WARNED', Color.RED)

        # Valid post
        else:
            log('--VALID', Color.GREEN)

            # Add to valid posts cache
            valid_posts.append(post_id)
            log('----caching', Color.GREEN)

            # Remove warning comment if it's valid now
            if post_id in already_warned:
                log('--Removing any warnings')
                comments_query = "SELECT fromid, id FROM comment" + \
                                 " WHERE post_id=\"" + str(post_id) + "\""
                comments = graph.fql(comments_query)
                for comment in comments:
                    if comment['fromid'] == bot_id:
                        # Delete warning comment
                        graph.delete_object(comment['id'])
                        log('--Warning deleted')

                        # Message the user notifying them the comment is deleted
                        # and thank them for fixing their post
                        log('--Thanking user')
                        send_message(str(actor_id), str(bot_id),
                                     "Thank you for fixing your post," +
                                     " I have removed the warning comment.",
                                     str(sublets_api_id),
                                     str(sublets_oauth_access_token))

                # Remove post from list of warned people
                log('--Removing from cache')
                del already_warned[post_id]

    # Save the updated caches
    log('Saving warned cache', Color.BOLD)
    save_cache(warned_db, already_warned)

    log('Saving valid cache', Color.BOLD)
    save_cache(valid_db, valid_posts)

    save_properties(saved_props)

    # Done
    notify_mac()


# Main method
if __name__ == "__main__":
    # Check to see if we're running on Heroku
    if os.environ.get('MEMCACHEDCLOUD_SERVERS', None):
        import bmemcached

        log('Running on heroku, using memcached', Color.BOLD)

        # Authenticate Memcached
        running_on_heroku = True
        mc = bmemcached.Client(os.environ.get('MEMCACHEDCLOUD_SERVERS').
                               split(','),
                               os.environ.get('MEMCACHEDCLOUD_USERNAME'),
                               os.environ.get('MEMCACHEDCLOUD_PASSWORD'))

    args = sys.argv
    # parser = argparse.ArgumentParser()
    # parser.add_argument("-e", help="extend the access token on run")
    # parser.add_argument("-i", help="initialize properties")
    # parser.add_argument("-t", help="only run test method")
    # parser.add_argument("-p", help="set some property values")
    # args = parser.parse_args()

    # Arg parsing. I know, there's better ways to do this
    if len(args) > 1:
        if "--extend" in args:
            extend_key = True
        elif "setprops" in args:
            set_new_props()
            sys.exit()
        elif "init" in args:
            init_props()
            sys.exit()
        elif "test" in args:
            test()
            sys.exit("Done testing")
        else:
            sys.exit('No valid args specified')

    sub_group()
